virtctl console wait until the virtual machine is in running state

SchSeba opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
virtctl console wait until the virtual machine is in running state

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1001 

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
virtctl console wait until the virtual machine is in running state
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubevirt/kubevirt/1323)
<!-- Reviewable:end -->

booxter commented on pkg/virtctl/console/console.go:
```diff
@@ -93,19 +93,53 @@ func (c *Console) Run(cmd *cobra.Command, args []string) error {
 
+	// Wait until the virtual machine is in running phase, user interrupt or timeout
+	runningChan := make(chan error)
+	timeoutChan := time.Tick(time.Duration(timeout) * time.Minute)
```
this ticker goroutine never stops though we don't need it running while interacting with the console. I think we should switch to time.NewTicker and explicitly Stop() it when it's no longer needed.

SchSeba replied:
Done

booxter commented on pkg/virtctl/console/console.go:
```diff
@@ -93,19 +93,53 @@ func (c *Console) Run(cmd *cobra.Command, args []string) error {
 
+	select {
+	case <-timeoutChan:
+		return fmt.Errorf("Timeout trying to connect to the virtual machine instance")
```
there is a chance now that a timeout happens while the goroutine above actually succeeds to connect. Is there anything to clean up in this particular case? Particularly, doesn't con.Stream hold a connection state that would be better off closed with explicit close() call?

SchSeba replied:
Because this is a command line program if the timeout or the interrupt channel is active the program will exit so the connection will be closed anyway.

booxter replied:
Do you mean some deferred callback handles it for us? If so, great.

SchSeba commented:
> Nice work. One thing to improve beyond the code is commit message. It currently has a very long title that wraps. There is a common guideline about titles being quite short, and descriptions expanding on the titles. For example, see: https://chris.beams.io/posts/git-commit/

Sure no problem I will make the tittle shorter.

> It would also make sense to refer to the issue you are solving in the commit messages. For that, you may refer to https://help.github.com/articles/closing-issues-using-keywords/ It's just putting "Closes #..." in the commit message. Then whenever we merge the PR, the issue will be closed. (And also, references between PR and the issue will be established.)

About the closes our bot is making the same thing by adding the `Fixes #1001` in the PR

SchSeba commented on pkg/virtctl/console/console.go:
```diff
@@ -93,19 +93,53 @@ func (c *Console) Run(cmd *cobra.Command, args []string) error {
-		if err != nil {
-			resChan <- err
-			return
+		if strings.Contains(fmt.Sprint(err), "not found") {
```
Done

SchSeba commented on pkg/virtctl/console/console.go:
```diff
@@ -93,19 +93,53 @@ func (c *Console) Run(cmd *cobra.Command, args []string) error {
+		for err != nil {
+			// Sleep to prevent denial of service on the api server
+			time.Sleep(500 * time.Millisecond)
+			con, err = virtCli.VirtualMachineInstance(namespace).SerialConsole(vmi)
```
I changed the check now its inside the for loop.

stu-gott commented on pkg/virtctl/console/console.go:
```diff
@@ -93,19 +93,60 @@ func (c *Console) Run(cmd *cobra.Command, args []string) error {
-		if err != nil {
-			resChan <- err
-			return
+		for err != nil {
```
Sorry for the tardy review. I thought I had reviewed yesterday but my comment seems to have disappeared.

It might be worth looking into the "k8s.io/apimachinery/pkg/util/wait" library. There's a PollImmediate function that already handles the busywork of polling on an interval with a timeout/error conditions. Using that could possibly simplify this code a bit.

SchSeba replied:
I take a look into the function and make the implementation but this not simplify the code.
This help with the timeout but not with the user interrupt.
Also we need to create a new goroutine after the PollImmediate success to create a new connection to the console service for the connection stream.

booxter commented:
The latest code changes are spot on, thanks (I would split the ticker change from error type change into separate commits since they are not both related to the commit message of the commit they are lumped into), but that's probably a minor issue.

The commit message title for one of the commits is still too long so it would be nice to see this fixed. Also to clarify, the reason why we may want to mention the issue number in PR commit messages and not just PR description on github.com is that in the former case, we have a direct link to the issue; while in the latter case, there are several levels of indirection (the reader needs to first identify which merge commit introduced a commit in the master tree; then follow the PR number from the merge commit message to the PR; then finally follow the link to the issue). [It's not a major thing, but may alleviate some of the pain in the future.]

One thing that I've missed in the previous round of review, and I am sorry for not pointing it out right away, is that the PR lacks new tests for the change. I think it would benefit from a functional test case that would start a VMI and immediately call to console and check that the tool actually connects to the VMI (after some time when the guest is up and ready) and not bails out on the user with an error, as it did before the PR. I haven't checked if we already have functional test cases for the CLI, but if not, that's an invitation to add some.

Thanks.

SchSeba commented:
@booxter 
> The commit message title for one of the commits is still too long so it would be nice to see this fixed.
Changed.

> One thing that I've missed in the previous round of review, and I am sorry for not pointing it out right away, is that the PR lacks new tests for the change. I think it would benefit from a functional test case that would start a VMI and immediately call to console and check that the tool actually connects to the VMI (after some time when the guest is up and ready) and not bails out on the user with an error, as it did before the PR. I haven't checked if we already have functional test cases for the CLI, but if not, that's an invitation to add some.

I Changed how the console functional tests works by removing the waitForVMIStart call before we call the console.

We don't have functional test cases for the CLI and I am not sure how to test the console because if I call the console command from the code a get an error that he cant find a `tty` related to this line `terminal.MakeRaw(int(os.Stdin.Fd()))`

booxter commented on tests/utils.go:
```diff
@@ -1130,10 +1131,21 @@ func NewConsoleExpecter(virtCli kubecli.KubevirtClient, vmi *v1.VirtualMachineIn
-		if err != nil {
-			resCh <- err
-			return
+		con, err := virtCli.VirtualMachineInstance(vmi.Namespace).SerialConsole(vmi.Name)
```
I think the idea of making functional tests implicitly validate console waiting for scheduler is great and better than having one or two cases to cover this behavior explicitly.

But it doesn't seem that you really validate the change in the tool here. This code below is a duplicate of what is added in the tool; you don't reuse the code from there. I think the right thing to do is to reuse the code from the tool here. It may require spinning up some of the tool code into a separate helper function ready to use.

As-is, I suspect the tests changes would pass even without the fix in the tool.

booxter commented:
I understand that if you just call the tool from inside tests directly then you get `terminal.MakeRaw(int(os.Stdin.Fd()))` called, which would be wrong. I think it suggests that you may need to spin up the goroutine, or its part, that establishes connection to the console (or fails when the timeout is reached or the VMI is not found) into a separate function that you would be able to call to from both the tool `main` body as well as from `tests/utils.go`.

stu-gott commented:
retest this please

booxter commented on pkg/kubecli/vmi.go:
```diff
@@ -246,7 +247,21 @@ func (v *vmis) VNC(name string) (StreamInterface, error) {
 func (v *vmis) SerialConsole(name string) (StreamInterface, error) {
-	return v.asyncSubresourceHelper(name, "console")
+	con, err := v.asyncSubresourceHelper(name, "console")
+	for err != nil {
```
Isn't it an issue that the function may never exit? The change may hang functional tests indefinitely, among other things.

cynepco3hahue commented:
@SchSeba you need to use affinity from k8s API to avoid
```
# kubevirt.io/kubevirt/tests_test
tests/console_test.go:113:26: undefined: "kubevirt.io/kubevirt/pkg/api/v1".Affinity
make: *** [goveralls] Error 2
```

SchSeba commented:
Look like there is a problem with the travis-ci.
Complain on `git status --porcelain` and `tests/console_test.go:113:26: undefined: "kubevirt.io/kubevirt/pkg/api/v1".Affinity` but the tests runs without any problem on my local environment and my git status is empty.

cynepco3hahue commented:
@SchSeba Rebase on master I saw Roman merged some change to use k8s `Affinity` instead of creating our own.

SchSeba commented:
Rebase do the magic thanks @cynepco3hahue :)

booxter commented on pkg/kubecli/vmi.go:
```diff
@@ -245,8 +246,55 @@ func (ws *wsStreamer) Stream(options StreamOptions) error {
+
+	select {
+	case <-timeoutChan:
+		return nil, fmt.Errorf("Timeout trying to connect to the virtual machine instance")
```
it's nice that we eventually exit but shouldn't we also shut down the goroutine?

SchSeba replied:
I don't think we need to stop the goroutine because we have 3 scenarios:
1. the helper return the con or error i will exit the goroutine
2. timeout this return an error an the virctl exit
3. user interrupt the virtctl exit

all of them will stop the goroutine.

booxter replied:
And when it's used in a functional test, do you terminate the goroutine (before the whole suite exits)? I don't think it's generate a good idea to leak it to outside even if the caller will eventually exit.

booxter commented on tests/utils.go:
```diff
@@ -1130,7 +1130,7 @@ func NewConsoleExpecter(virtCli kubecli.KubevirtClient, vmi *v1.VirtualMachineIn
 	resCh := make(chan error)
 	go func() {
-		con, err := virtCli.VirtualMachineInstance(vmi.ObjectMeta.Namespace).SerialConsole(vmi.ObjectMeta.Name)
+		con, err := virtCli.VirtualMachineInstance(vmi.Namespace).SerialConsole(vmi.Name, int(timeout))
```
I think now that you reuse the same timeout both to wait for serial console to start AND later for the expecter, you should track how much time it took to connect to the serial console, and reduce the timeout that is used for expecter by that much. Otherwise this function may take significantly longer than expected by the caller, in case when it took significant time for serial console to connect.

SchSeba replied:
if it took a lot of time to connect the test will exit because of the test time out no?

I don't think we need to make this complex because the test have timeout for all the test scenario (create vmi, wait for console, wait for the expecter).

booxter replied:
Yes but your timeout semantics is now broken (in some cases): the function receives a timeout argument but then doesn't satisfy it (by taking up to twice the expected time). While indeed it probably works (=passes tests) in most cases, it's not ideal.

booxter commented on tests/utils.go:
```diff
@@ -1129,13 +1129,15 @@ func NewConsoleExpecter(virtCli kubecli.KubevirtClient, vmi *v1.VirtualMachineIn
-			return
-		}
 
+	startTime := time.Now()
```
Nice!

booxter commented:
@vladikr I think this PR is good to go; I am not sure about CI failures. I tried to deduce what failed in the jobs but couldn't find a place to extract more detailed logs for the failures. Are these failures known?

vladikr commented:
ok to test

vladikr commented:
not sure why kubevirt functional tests didn't run.

SchSeba commented:
@booxter @vladikr All the tests looks good now.

booxter commented:
I don't have full exposure as to which jobs are supposed to pass, but I see that 1.10.3 all pass (except openshift that is expected to fail at least because of custom MAC address issue; and other failures don't seem related either). I will take it upon myself to merge this patch. I hope that's ok.

rmohr commented on pkg/kubecli/vmi.go:
```diff
@@ -245,8 +246,59 @@ func (ws *wsStreamer) Stream(options StreamOptions) error {
+			if asyncSubresourceError, ok := err.(*AsyncSubresourceError); ok {
+				if asyncSubresourceError.GetStatusCode() == http.StatusBadRequest {
+					// Sleep to prevent denial of service on the api server
+					time.Sleep(1 * time.Second)
```
I think it is better to use watches instead of polling. That should be much cheaper once the connection is established.
